### PR TITLE
[Feature] Add support for multiple oracle sources per token in Scope program

### PR DESCRIPTION
**Pull request by oselezi**
**Overview:**
Added a feature by allowing each token to have multiple oracle sources. This adds redundancy and improves reliability by enabling the protocol to use alternative price feeds if one fails or provides invalid data. This actually happening recently in some Solana projects.

**Changes:**
Updated the `OracleMappings` struct to support multiple oracle accounts per token.
Modified handlers to accommodate the new structure, including `handler_update_mapping.rs` and `handler_refresh_prices.rs`.
Adjusted methods and logic to iterate over multiple sources and select valid prices.
Added error handling for invalid source indices.
Adjusted account sizes.

**Testing:**
Added unit tests for the new functionality. (on local)
Performed integration tests to verify correct behavior.

### CODE REVIEW

**Comment by oselezi** [comment]
@oeble

File: programs/scope/src/states.rs [hunk 1]
```diff
@@ -115,26 +117,31 @@ pub struct OraclePrices {
...
-    pub twap_enabled: [u8; MAX_ENTRIES], // true or false
+    // Each token can have up to MAX_SOURCES sources
+    pub price_info_accounts: [[Pubkey; MAX_SOURCES]; MAX_ENTRIES],
+    pub price_types: [[u8; MAX_SOURCES]; MAX_ENTRIES],
+    pub twap_source: [[u16; MAX_SOURCES]; MAX_ENTRIES], // meaningful only if type == TWAP; the index of where we find the TWAP
+    pub twap_enabled: [[u8; MAX_SOURCES]; MAX_ENTRIES], // true or false
     pub ref_price: [u16; MAX_ENTRIES], // reference price against which we check confidence within 5%
-    pub generic: [[u8; 20]; MAX_ENTRIES], // generic data parsed depending on oracle type
+    pub generic: [[[u8; 20]; MAX_SOURCES]; MAX_ENTRIES], // generic data parsed depending on oracle type
 }
```
Comment (oeble) [inline programs/scope/src/states.rs:136]: This modifies the states in a way that makes it incompatible with the current version. This program is deployed and in use, and we can't introduce this change in the way you suggest.

**Review by oeble — COMMENTED** [review]
@oselezi Thank you for your submission. Supporting multiple sources for a price would be a nice addition, I cannot accept this PR.

1. This PR modifies the states in a way that makes it incompatible with the current version. This program is deployed and in use, and we can't introduce this change in the way you suggest.
2. When we will introduce a feature similar to this one in scope, we want a more complex way of defining the rules to elect or compute the final price from multiple PRs.

File: programs/scope/src/states.rs (hunk 1 above)
Comment (oselezi) [reply programs/scope/src/states.rs:136]: @oeble Thank you for reviewing my PR and for your feedback. I understand the concerns regarding state incompatibility with the current deployed version of the program.
I'm interested in exploring ways to implement support for multiple oracle sources per token in a backward-compatible manner. I also recognize that a more sophisticated approach to price aggregation is desired.
Could you provide more details on how you envision the rules for electing or computing the final price from multiple sources? I'd be happy to collaborate and adjust my implementation to align with the project's roadmap and requirements. I am reviewing other repositories as well, if you need any part that need to be reviewed most, please let me know. Thanks

Comment (oselezi) [reply programs/scope/src/states.rs:136]: Closing this PR, I will move to other solution

**Pull request closed without merging**